homeassistant/components/isy994.py

```diff
@@ -44,190 +43,352 @@
+                        'RelayLampOnly_ADV', 'KeypadButton',
+                        'KeypadButton_ADV', 'EZRAIN_Input', 'EZRAIN_Output',
+                        'EZIO2x4_Input', 'EZIO2x4_Input_ADV', 'BinaryControl',
+                        'BinaryControl_ADV', 'AlertModuleSiren',
+                        'AlertModuleSiren_ADV', 'AlertModuleArmed', 'Siren',
+                        'Siren_ADV'],
+        'insteon_type': ['2.', '9.10.', '9.11.']
+    }
+}
+
+SUPPORTED_DOMAINS = ['binary_sensor', 'sensor', 'lock', 'fan', 'cover',
+                     'light', 'switch']
+SUPPORTED_PROGRAM_DOMAINS = ['binary_sensor', 'lock', 'fan', 'cover', 'switch']
+
+# ISY Scenes are more like Swithes than Hass Scenes
+# (they can turn off, and report their state)
+SCENE_DOMAIN = 'switch'
+
+ISY994_NODES = "isy994_nodes"
+ISY994_WEATHER = "isy994_weather"
+ISY994_PROGRAMS = "isy994_programs"
 
-PYISY = None
+WeatherNode = namedtuple('WeatherNode', ('status', 'name', 'uom'))
 
-HIDDEN_STRING = DEFAULT_HIDDEN_STRING
 
-SUPPORTED_DOMAINS = ['binary_sensor', 'cover', 'fan', 'light', 'lock',
-                     'sensor', 'switch']
+def _check_for_node_def(hass: HomeAssistant, node,
+                        single_domain: str=None) -> bool:
+    """Check if the node matches the node_def_id for any domains.
 
+    This is only present on the 5.0 ISY firmware, and is the most reliable
+    way to determine a device's type.
+    """
+    try:
+        node_def_id = node.node_def_id
+    except AttributeError:
+        # Node doesn't have a node_def
+        return False
 
-WeatherNode = namedtuple('WeatherNode', ('status', 'name', 'uom'))
+    domains = SUPPORTED_DOMAINS if not single_domain else [single_domain]
+    for domain in domains:
+        if node_def_id in NODE_FILTERS[domain]['node_def_id']:
+            hass.data[ISY994_NODES][domain].append(node)
+            return True
 
+    return False
 
-def filter_nodes(nodes: list, units: list=None, states: list=None) -> list:
-    """Filter a list of ISY nodes based on the units and states provided."""
-    filtered_nodes = []
-    units = units if units else []
-    states = states if states else []
-    for node in nodes:
-        match_unit = False
-        match_state = True
-        for uom in node.uom:
-            if uom in units:
-                match_unit = True
-                continue
-            elif uom not in states:
-                match_state = False
 
-            if match_unit:
-                continue
+def _check_for_insteon_type(hass: HomeAssistant, node,
+                            single_domain: str=None) -> bool:
+    """Check if the node matches the Insteon type for any domains.
 
-        if match_unit or match_state:
-            filtered_nodes.append(node)
+    This is for (presumably) every version of the ISY firmware, but only
+    works for Insteon device. "Node Server" (v5+) and Z-Wave and others will
+    not have a type.
+    """
+    try:
+        device_type = node.type
```

Comment:
instead to make a exception and hard interrupt the flow of your program, just check if type will be a attribute.